pkg/segment/query/iqr/intermediateQueryResult.go

```diff
@@ -716,17 +718,162 @@ func (iqr *IQR) AsResult() (*structs.PipeSearchResponseOuter, error) {
+			Qtype:              iqr.qType.String(),
+			CanScrollMore:      false,
+			ColumnsOrder:       toputils.GetSortedStringKeys(records),
+			BucketCount:        bucketCount,
+			MeasureFunctions:   measureFuncs,
+			MeasureResults:     bucketHolderArr,
+			GroupByCols:        aggGroupByCols,
+			// TODO: add IsTimechart flag
+		}
+	case structs.SegmentStatsCmd: // TODO: implement
+		panic("IQR.AsResult: SegmentStatsCmd not implemented")
+	default:
+		return nil, fmt.Errorf("IQR.AsResult: unexpected query type %v", iqr.qType)
 	}
 
 	return response, nil
 }
+
+func (iqr *IQR) AppendGroupByResults(bucketHolderArr []*structs.BucketHolder, measureFuncs []string, aggGroupByCols []string, bucketCount int) error {
+	if err := iqr.validate(); err != nil {
+		log.Errorf("IQR.AppendGroupByResults: validation failed: %v", err)
+		return err
+	}
+
+	if iqr.qType != structs.GroupByCmd {
+		return fmt.Errorf("IQR.AppendGroupByResults: not a GroupBy query. queryType=%v", iqr.qType)
+	}
+
+	iqr.mode = withoutRRCs
+
+	knownValues := make(map[string][]utils.CValueEnclosure)
+
+	for _, aggGroupByCol := range aggGroupByCols {
+		knownValues[aggGroupByCol] = make([]utils.CValueEnclosure, bucketCount)
+	}
+	for _, retMFun := range measureFuncs {
+		knownValues[retMFun] = make([]utils.CValueEnclosure, bucketCount)
+	}
+
+	conversionErrors := make([]string, utils.MAX_SIMILAR_ERRORS_TO_LOG)
+	errIndex := 0
+
+	for i, bucketHolder := range bucketHolderArr {
+		for idx, aggGroupByCol := range aggGroupByCols {
+			colValue := bucketHolder.GroupByValues[idx]
+			err := knownValues[aggGroupByCol][i].ConvertValue(colValue)
+			if err != nil && errIndex < utils.MAX_SIMILAR_ERRORS_TO_LOG {
+				conversionErrors[i] = fmt.Sprintf("BucketHolderIndex=%v, groupByCol=%v, ColumnValue=%v", i, aggGroupByCol, colValue)
+				errIndex++
+			}
+		}
+
+		for _, measureFunc := range measureFuncs {
+			value := bucketHolder.MeasureVal[measureFunc]
+			err := knownValues[measureFunc][i].ConvertValue(value)
+			if err != nil && errIndex < utils.MAX_SIMILAR_ERRORS_TO_LOG {
+				conversionErrors[i] = fmt.Sprintf("BucketHolderIndex=%v, measureFunc=%v, ColumnValue=%v", i, measureFunc, value)
+				errIndex++
+			}
+		}
+	}
+
+	err := iqr.AppendKnownValues(knownValues)
+	if err != nil {
+		return err
+	}
+
+	iqr.groupbyColumns = append(iqr.groupbyColumns, aggGroupByCols...)
+	iqr.measureColumns = append(iqr.measureColumns, measureFuncs...)
+
+	if errIndex > 0 {
+		log.Errorf("IQR.AppendGroupByResults: conversion errors: %v", conversionErrors)
+	}
+
+	return nil
+}
+
+func (iqr *IQR) getFinalResultForGroupBy() ([]*structs.BucketHolder, []string, []string, int, error) {
+	knownValues := iqr.knownValues
+
+	if len(knownValues) == 0 {
+		return nil, nil, nil, 0, fmt.Errorf("IQR.getFinalResultForGroupBy: knownValues is empty")
+	}
+
+	// The bucket count is the number of rows in the final result. So we can use the length of any column.
+	bucketCount := len(knownValues[iqr.groupbyColumns[0]])
```

Comment:
You might want to check the length of `iqr.groupbyColumns` first and return an error if it's 0, so we don't panic if we ever get to that state

Reply:
Handled this in the PR: #1774